File: agents/python/src/tools/execute_command.py
"""Tool: execute_command — run shell commands with timeout."""

import subprocess
from typing import Any


def execute_command(args: dict[str, Any], project_path: str) -> str:
    command = args["command"]
    timeout = args.get("timeout", 30000) / 1000  # Convert ms to seconds

    try:
        result = subprocess.run(
            command,
            shell=True,
            cwd=project_path,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        return (
            f"Exit code: {result.returncode}\n"
            f"Stdout:\n{result.stdout}\n"
            f"Stderr:\n{result.stderr}"
        )
    except subprocess.TimeoutExpired:
        return f"Exit code: -1\nError: Command timed out after {timeout}s"
    except Exception as e:
        return f"Exit code: -1\nError: {e}"

File: agents/rust/src/tools/execute_command.rs
use std::collections::HashMap;
use std::process::Command;
use std::time::Duration;

pub fn execute_command(args: &HashMap<String, serde_json::Value>, project_path: &str) -> String {
    let command = args.get("command").and_then(|v| v.as_str()).unwrap_or("");
    let _timeout_ms = args.get("timeout").and_then(|v| v.as_u64()).unwrap_or(30000);

    let output = Command::new("sh")
        .arg("-c")
        .arg(command)
        .current_dir(project_path)
        .output();

    match output {
        Ok(output) => {
            let exit_code = output.status.code().unwrap_or(-1);
            let stdout = String::from_utf8_lossy(&output.stdout);
            let stderr = String::from_utf8_lossy(&output.stderr);
            format!("Exit code: {}\nStdout:\n{}\nStderr:\n{}", exit_code, stdout, stderr)
        }
        Err(e) => format!("Exit code: -1\nError: {}", e),
    }
}
